File: LeetCode-Solved-Problems/1608-special-array-with-x-elements-greater-than-or-equal-x/1608-special-array-with-x-elements-greater-than-or-equal-x.py
class Solution:
    def specialArray(self, nums: List[int]) -> int:
        nums.sort()
        i = 0
        prev = -1
        total_right = len(nums)
        
        while i < len(nums):
            if nums[i] == total_right or (prev < total_right < nums[i]):
                return total_right
            
            while i+1 < len(nums) and nums[i] == nums[i+1]:
                i += 1
            
            prev = nums[i]
            i+=1
            total_right = len(nums) - i
            
        return -1

File: LeetCode-Solved-Problems/1608-special-array-with-x-elements-greater-than-or-equal-x/1608-special-array-with-x-elements-greater-than-or-equal-x.cpp
class Solution {
public:
    int specialArray(vector<int>& nums) {
        // Time: O(n), Space: O(n)
        int n = nums.size();
        vector<int> count(n+1, 0);
        for(int num:nums){
            int indx = num < n? num:n;
            count[indx]++;
        }
        
        int total_right = 0;
        for(int i = n; i>=0; i--){
            total_right+=count[i];
            if(i == total_right)
                return total_right;
        }
        
        return -1;
    }
};